File: src/main/java/com/sfh/pokeRogueBot/model/modifier/impl/PokemonHpRestoreModifierItem.java
package com.sfh.pokeRogueBot.model.modifier.impl;

import com.sfh.pokeRogueBot.model.poke.Pokemon;
import lombok.Data;
import lombok.EqualsAndHashCode;

@Data
@EqualsAndHashCode(callSuper = true)
public class PokemonHpRestoreModifierItem extends PokemonModifierItem {

    public static final String TARGET = "PokemonHpRestoreModifierType";

    private boolean healStatus;
    private int restorePercent;
    private int restorePoints;

    public void apply(Pokemon target) {
        int hp = target.getHp();
        int hpPointsRestored = hp + restorePoints;
        int hpPercentRestored = hp + (int) (target.getStats().getHp() * (restorePercent / 100.0));

        int newHp = Math.max(hpPointsRestored, hpPercentRestored);
        if (newHp > target.getStats().getHp()) {
            newHp = target.getStats().getHp();
        }

        target.setHp(newHp);
    }
}

File: src/main/kotlin/com/sfh/pokeRogueBot/model/modifier/impl/PokemonHpRestoreModifierItem.kt
package com.sfh.pokeRogueBot.model.modifier.impl

import com.sfh.pokeRogueBot.model.poke.Pokemon
import kotlin.math.max

open class PokemonHpRestoreModifierItem : PokemonModifierItem() {

    companion object {
        const val TARGET = "PokemonHpRestoreModifierType"
    }

    var healStatus: Boolean = false
    var restorePercent: Int = 0
    var restorePoints: Int = 0

    fun apply(target: Pokemon) {
        val hp = target.hp
        val hpPointsRestored = hp + restorePoints
        val hpPercentRestored = hp + (target.stats.hp * (restorePercent / 100.0)).toInt()

        var newHp = max(hpPointsRestored, hpPercentRestored)
        if (newHp > target.stats.hp) {
            newHp = target.stats.hp
        }

        target.hp = newHp
    }
}